src/v5/types/observablearray.ts

```diff
@@ -474,27 +472,48 @@ const arrayExtensions = {
      * Without this, everything works as well, but this works
      * faster as everything works on unproxied values
      */
+;["concat", "indexOf", "join", "lastIndexOf", "slice", "toString", "toLocaleString"].forEach(
+    funcName => {
+        arrayExtensions[funcName] = function() {
+            const adm: ObservableArrayAdministration = this[$mobx]
+            adm.atom.reportObserved()
+            const res = adm.dehanceValues(adm.values)
+            return res[funcName].apply(res, arguments)
+        }
+    }
+)
+
+/**
+ * Make sure callbacks recieve correct array arg #2326
+ * Impl assumes that callback's last 3 arguments are:
+ * element, index, array
+ */
 ;[
-    "concat",
     "every",
     "filter",
+    "find",
+    "findIndex",
+    "flatMap",
     "forEach",
-    "indexOf",
-    "join",
-    "lastIndexOf",
     "map",
     "reduce",
     "reduceRight",
-    "slice",
-    "some",
-    "toString",
-    "toLocaleString"
+    "some"
 ].forEach(funcName => {
-    arrayExtensions[funcName] = function() {
+    // Feature detection (eg flatMap may not be available)
+    if (typeof Array.prototype[funcName] !== "function") {
+        return
+    }
+    arrayExtensions[funcName] = function(callback, thisArg) {
         const adm: ObservableArrayAdministration = this[$mobx]
         adm.atom.reportObserved()
-        const res = adm.dehanceValues(adm.values)
-        return res[funcName].apply(res, arguments)
+        return adm.values[funcName]((...args) => {
+            // replace last arg (the array) with observable array
+            args[args.length - 1] = this
+            // replace last but two arg (the element) with dehanced value
+            args[args.length - 3] = adm.dehanceValue(args[args.length - 3])
+            return callback.apply(thisArg, args)
```

Comment:
very subtle issue here that for `reduce` and `reduceRight` the context will always be the `initialValue` instead of `null`/`global` .